# 导入库

In [4]:
import time  
import numpy as np  
import pandas as pd 
from sklearn.ensemble import RandomForestClassifier
# 随机森林分类器
# ensemble乐团；整体；全体；全套服装
from pyecharts import Bar3D 
# 3D柱形图 RFM模型使用
# pip install pyecharts==0.5.11

# 读取数据

In [5]:
road = r"D:\sales.xlsx"

In [6]:
# 按年份从数据库导出的一张excel表包含多张子表，及会员等级表
sheet_names = ['2015','2016','2017','2018','会员等级']
sheet_names

['2015', '2016', '2017', '2018', '会员等级']

In [7]:
sheet_datas = [pd.read_excel(road,sheet_name=i) for i in sheet_names]

In [8]:
type(sheet_datas)
# 列表数据类型，所有数据行，包括列名
# 均保存到list内

list

In [9]:
sheet_datas[0:2]
# 查看从第0张表开始的两张 0表和1表

[              会员ID         订单号       提交日期    订单金额
 0      15278002468  3000304681 2015-01-01   499.0
 1      39236378972  3000305791 2015-01-01  2588.0
 2      38722039578  3000641787 2015-01-01   498.0
 3      11049640063  3000798913 2015-01-01  1572.0
 4      35038752292  3000821546 2015-01-01    10.1
 ...            ...         ...        ...     ...
 30769  39368100847  4281994827 2015-12-31   828.0
 30770       409757  4282010457 2015-12-31   199.0
 30771  38380526114  4282017675 2015-12-31   208.0
 30772     28074988  4282019440 2015-12-31    89.0
 30773  39460363230  4282025309 2015-12-31   719.0
 
 [30774 rows x 4 columns],               会员ID         订单号       提交日期     订单金额
 0      39288120141  4282025766 2016-01-01    76.00
 1      39293812118  4282037929 2016-01-01  7599.00
 2      27596340905  4282038740 2016-01-01   802.00
 3      15111475509  4282043819 2016-01-01    65.00
 4      38896594001  4282051044 2016-01-01    95.00
 ...            ...         ...        ...      

In [78]:
for i in sheet_names:
    print(i)
    print(type(i))

2015
<class 'str'>
2016
<class 'str'>
2017
<class 'str'>
2018
<class 'str'>
会员等级
<class 'str'>


In [81]:
a = pd.read_excel(road,sheet_name='2015')
a

会员ID         订单号       提交日期    订单金额
0      15278002468  3000304681 2015-01-01   499.0
1      39236378972  3000305791 2015-01-01  2588.0
2      38722039578  3000641787 2015-01-01   498.0
3      11049640063  3000798913 2015-01-01  1572.0
4      35038752292  3000821546 2015-01-01    10.1
...            ...         ...        ...     ...
30769  39368100847  4281994827 2015-12-31   828.0
30770       409757  4282010457 2015-12-31   199.0
30771  38380526114  4282017675 2015-12-31   208.0
30772     28074988  4282019440 2015-12-31    89.0
30773  39460363230  4282025309 2015-12-31   719.0

[30774 rows x 4 columns]

In [82]:
b = pd.read_excel(road,sheet_name='2017')
b

会员ID         订单号       提交日期    订单金额
0      38765290840  4324911135 2017-01-01  1799.0
1      39305832102  4324911213 2017-01-01   369.0
2      34190994969  4324911251 2017-01-01   189.0
3      38986333210  4324911283 2017-01-01   169.0
4          4271359  4324911355 2017-01-01    78.0
...            ...         ...        ...     ...
50834  39155833075  4338762307 2017-12-31    96.0
50835  29523124076  4338762514 2017-12-31    99.0
50836      7807573  4338762906 2017-12-31   258.0
50837  36642530033  4338762958 2017-12-31   286.0
50838  39155901156  4338763879 2017-12-31   249.0

[50839 rows x 4 columns]

In [83]:
a = a.append(b)
a.head(-1)

会员ID         订单号       提交日期    订单金额
0      15278002468  3000304681 2015-01-01   499.0
1      39236378972  3000305791 2015-01-01  2588.0
2      38722039578  3000641787 2015-01-01   498.0
3      11049640063  3000798913 2015-01-01  1572.0
4      35038752292  3000821546 2015-01-01    10.1
...            ...         ...        ...     ...
50833  39304933721  4338761879 2017-12-31  1599.0
50834  39155833075  4338762307 2017-12-31    96.0
50835  29523124076  4338762514 2017-12-31    99.0
50836      7807573  4338762906 2017-12-31   258.0
50837  36642530033  4338762958 2017-12-31   286.0

[81612 rows x 4 columns]

In [84]:
a = a.reset_index()
a.head(-1)

index         会员ID         订单号       提交日期    订单金额
0          0  15278002468  3000304681 2015-01-01   499.0
1          1  39236378972  3000305791 2015-01-01  2588.0
2          2  38722039578  3000641787 2015-01-01   498.0
3          3  11049640063  3000798913 2015-01-01  1572.0
4          4  35038752292  3000821546 2015-01-01    10.1
...      ...          ...         ...        ...     ...
81607  50833  39304933721  4338761879 2017-12-31  1599.0
81608  50834  39155833075  4338762307 2017-12-31    96.0
81609  50835  29523124076  4338762514 2017-12-31    99.0
81610  50836      7807573  4338762906 2017-12-31   258.0
81611  50837  36642530033  4338762958 2017-12-31   286.0

[81612 rows x 5 columns]

In [85]:
a.订单金额 == 249.0

0        False
1        False
2        False
3        False
4        False
         ...  
81608    False
81609    False
81610    False
81611    False
81612     True
Name: 订单金额, Length: 81613, dtype: bool

In [9]:
sheet_datas[1]

会员ID         订单号       提交日期     订单金额
0      39288120141  4282025766 2016-01-01    76.00
1      39293812118  4282037929 2016-01-01  7599.00
2      27596340905  4282038740 2016-01-01   802.00
3      15111475509  4282043819 2016-01-01    65.00
4      38896594001  4282051044 2016-01-01    95.00
...            ...         ...        ...      ...
41273  35336052906  4324910145 2016-12-31    99.00
41274  39305835721  4324910148 2016-12-31   238.89
41275  39296945352  4324910770 2016-12-31   765.00
41276  14791026234  4324911025 2016-12-31    45.80
41277  16779755770  4324911048 2016-12-31   119.00

[41278 rows x 4 columns]

# 数据描述-describe

In [10]:
sheet_names

['2015', '2016', '2017', '2018', '会员等级']

In [88]:
# sheet_datas.head(2)
# AttributeError: 'list' object has no attribute 'head'

sheet_datas[0:2]
# 每张表按行上下合并，包括了字段名称第一行

[              会员ID         订单号       提交日期    订单金额 max_year_date
 0      15278002468  3000304681 2015-01-01   499.0    2015-12-31
 1      39236378972  3000305791 2015-01-01  2588.0    2015-12-31
 2      38722039578  3000641787 2015-01-01   498.0    2015-12-31
 3      11049640063  3000798913 2015-01-01  1572.0    2015-12-31
 4      35038752292  3000821546 2015-01-01    10.1    2015-12-31
 ...            ...         ...        ...     ...           ...
 30769  39368100847  4281994827 2015-12-31   828.0    2015-12-31
 30770       409757  4282010457 2015-12-31   199.0    2015-12-31
 30771  38380526114  4282017675 2015-12-31   208.0    2015-12-31
 30772     28074988  4282019440 2015-12-31    89.0    2015-12-31
 30773  39460363230  4282025309 2015-12-31   719.0    2015-12-31
 
 [30574 rows x 5 columns],
               会员ID         订单号       提交日期     订单金额 max_year_date
 0      39288120141  4282025766 2016-01-01    76.00    2016-12-31
 1      39293812118  4282037929 2016-01-01  7599.00    2016

In [98]:
# TypeError: list indices must be integers or slices, not str
# list类型，非DataFrame类型无法单独调取2016的子表数据
# sheet_datas['2016'][0:2]

In [97]:
for i in sheet_datas:
    print(type(i))
    print(len(i))
    print(i.isnull().sum())
    print('~~~~~~~~~~~~~~~')

<class 'pandas.core.frame.DataFrame'>
30574
会员ID             0
订单号              0
提交日期             0
订单金额             0
max_year_date    0
dtype: int64
~~~~~~~~~~~~~~~
<class 'pandas.core.frame.DataFrame'>
41001
会员ID             0
订单号              0
提交日期             0
订单金额             0
max_year_date    0
dtype: int64
~~~~~~~~~~~~~~~
<class 'pandas.core.frame.DataFrame'>
50451
会员ID             0
订单号              0
提交日期             0
订单金额             0
max_year_date    0
dtype: int64
~~~~~~~~~~~~~~~
<class 'pandas.core.frame.DataFrame'>
80801
会员ID             0
订单号              0
提交日期             0
订单金额             0
max_year_date    0
dtype: int64
~~~~~~~~~~~~~~~
<class 'pandas.core.frame.DataFrame'>
154385
会员ID    0
会员等级    0
dtype: int64
~~~~~~~~~~~~~~~


In [10]:
for each_name,each_data in zip(sheet_names,sheet_datas):
    print('数据概要：{}'.format(each_name))
    print('数据摘要:',each_data.head())
    print('数据描述:',each_data.describe())
    print('缺失的情况',each_data.isnull().any(axis=1).sum())
    print('数据类型',each_data.dtypes)

数据概要：2015
数据摘要:           会员ID         订单号       提交日期    订单金额
0  15278002468  3000304681 2015-01-01   499.0
1  39236378972  3000305791 2015-01-01  2588.0
2  38722039578  3000641787 2015-01-01   498.0
3  11049640063  3000798913 2015-01-01  1572.0
4  35038752292  3000821546 2015-01-01    10.1
数据描述:                会员ID           订单号           订单金额
count  3.077400e+04  3.077400e+04   30774.000000
mean   2.918779e+10  4.020414e+09     960.991161
std    1.385333e+10  2.630510e+08    2068.107231
min    2.670000e+02  3.000305e+09       0.500000
25%    1.944122e+10  3.885510e+09      59.000000
50%    3.746545e+10  4.117491e+09     139.000000
75%    3.923593e+10  4.234882e+09     899.000000
max    3.954613e+10  4.282025e+09  111750.000000
缺失的情况 0
数据类型 会员ID             int64
订单号              int64
提交日期    datetime64[ns]
订单金额           float64
dtype: object
数据概要：2016
数据摘要:           会员ID         订单号       提交日期    订单金额
0  39288120141  4282025766 2016-01-01    76.0
1  39293812118  4282037929 2016-01

# 数据探索-异常-缺失-新特征

## 去掉子表-缺失和异常删除

In [102]:
len(sheet_datas)

5

In [101]:
len(sheet_datas[:-1])
# 最后一张表不加入到年份数据表

4

In [13]:
for i,each_data in enumerate(sheet_datas[:-1]): 
    print(i)

0
1
2
3


In [14]:
for i,each_data in enumerate(sheet_datas[:-1]): 
    print(i,each_data)

0               会员ID         订单号       提交日期    订单金额
0      15278002468  3000304681 2015-01-01   499.0
1      39236378972  3000305791 2015-01-01  2588.0
2      38722039578  3000641787 2015-01-01   498.0
3      11049640063  3000798913 2015-01-01  1572.0
4      35038752292  3000821546 2015-01-01    10.1
...            ...         ...        ...     ...
30769  39368100847  4281994827 2015-12-31   828.0
30770       409757  4282010457 2015-12-31   199.0
30771  38380526114  4282017675 2015-12-31   208.0
30772     28074988  4282019440 2015-12-31    89.0
30773  39460363230  4282025309 2015-12-31   719.0

[30774 rows x 4 columns]
1               会员ID         订单号       提交日期     订单金额
0      39288120141  4282025766 2016-01-01    76.00
1      39293812118  4282037929 2016-01-01  7599.00
2      27596340905  4282038740 2016-01-01   802.00
3      15111475509  4282043819 2016-01-01    65.00
4      38896594001  4282051044 2016-01-01    95.00
...            ...         ...        ...      ...
41273  353360

In [15]:
# 除去会员等级表单
for i,each_data in enumerate(sheet_datas[:-1]):   
    
    sheet_datas[i] = each_data.dropna()   
    # 删除缺失值记录
    
    sheet_datas[i] = each_data[each_data['订单金额']>1]  
    # 保留大于1元的记录
    
    sheet_datas[i]['max_year_date'] = each_data['提交日期'].max()

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  # Remove the CWD from sys.path while we load stuff.


In [16]:
sheet_datas[:-1][0].head(-1)
# 前4张年份表调取第一张后则为DataFrame类型可获取head数据

会员ID         订单号       提交日期    订单金额 max_year_date
0      15278002468  3000304681 2015-01-01   499.0    2015-12-31
1      39236378972  3000305791 2015-01-01  2588.0    2015-12-31
2      38722039578  3000641787 2015-01-01   498.0    2015-12-31
3      11049640063  3000798913 2015-01-01  1572.0    2015-12-31
4      35038752292  3000821546 2015-01-01    10.1    2015-12-31
...            ...         ...        ...     ...           ...
30768  39496685642  4281992204 2015-12-31   299.0    2015-12-31
30769  39368100847  4281994827 2015-12-31   828.0    2015-12-31
30770       409757  4282010457 2015-12-31   199.0    2015-12-31
30771  38380526114  4282017675 2015-12-31   208.0    2015-12-31
30772     28074988  4282019440 2015-12-31    89.0    2015-12-31

[30573 rows x 5 columns]

In [105]:
sheet_datas[:-1][2].head(-1)

会员ID         订单号       提交日期    订单金额 max_year_date
0      38765290840  4324911135 2017-01-01  1799.0    2017-12-31
1      39305832102  4324911213 2017-01-01   369.0    2017-12-31
2      34190994969  4324911251 2017-01-01   189.0    2017-12-31
3      38986333210  4324911283 2017-01-01   169.0    2017-12-31
4          4271359  4324911355 2017-01-01    78.0    2017-12-31
...            ...         ...        ...     ...           ...
50833  39304933721  4338761879 2017-12-31  1599.0    2017-12-31
50834  39155833075  4338762307 2017-12-31    96.0    2017-12-31
50835  29523124076  4338762514 2017-12-31    99.0    2017-12-31
50836      7807573  4338762906 2017-12-31   258.0    2017-12-31
50837  36642530033  4338762958 2017-12-31   286.0    2017-12-31

[50450 rows x 5 columns]

## 汇总所有数据

In [18]:
# 汇总所有数据（纵向汇总），原sheet_datas为list类型
# concate连接，axis0为每行的形式进行累加各个表
# 从15年到18年
data_merge = pd.concat(sheet_datas[:-1],axis=0)
data_merge.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 202827 entries, 0 to 81348
Data columns (total 5 columns):
会员ID             202827 non-null int64
订单号              202827 non-null int64
提交日期             202827 non-null datetime64[ns]
订单金额             202827 non-null float64
max_year_date    202827 non-null datetime64[ns]
dtypes: datetime64[ns](2), float64(1), int64(2)
memory usage: 9.3 MB


In [19]:
data_merge.head(2)

会员ID         订单号       提交日期    订单金额 max_year_date
0  15278002468  3000304681 2015-01-01   499.0    2015-12-31
1  39236378972  3000305791 2015-01-01  2588.0    2015-12-31

## 获取每行年份数据

In [20]:
# interval 间隔；区间；时间间隔
# 最大日期为订单在样本内的最后日期减去提交日期
# 则为每个行数据距离截止时间的间隔长度
data_merge['data_interval'] = data_merge['max_year_date'] - data_merge['提交日期']
data_merge.head(2)

会员ID         订单号       提交日期    订单金额 max_year_date data_interval
0  15278002468  3000304681 2015-01-01   499.0    2015-12-31      364 days
1  39236378972  3000305791 2015-01-01  2588.0    2015-12-31      364 days

In [21]:
data_merge['year'] = data_merge['提交日期'].dt.year
data_merge.head(2)

会员ID         订单号       提交日期    订单金额 max_year_date data_interval  \
0  15278002468  3000304681 2015-01-01   499.0    2015-12-31      364 days   
1  39236378972  3000305791 2015-01-01  2588.0    2015-12-31      364 days   

   year  
0  2015  
1  2015

In [22]:
data_merge.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 202827 entries, 0 to 81348
Data columns (total 7 columns):
会员ID             202827 non-null int64
订单号              202827 non-null int64
提交日期             202827 non-null datetime64[ns]
订单金额             202827 non-null float64
max_year_date    202827 non-null datetime64[ns]
data_interval    202827 non-null timedelta64[ns]
year             202827 non-null int64
dtypes: datetime64[ns](2), float64(1), int64(3), timedelta64[ns](1)
memory usage: 12.4 MB


## 转换日期间隔为数字

In [33]:
data_merge.head(2)['data_interval']

0   364 days
1   364 days
Name: data_interval, dtype: timedelta64[ns]

In [34]:
data_merge.head(2)['data_interval'][0]

Timedelta('364 days 00:00:00')

In [48]:
# data_merge.head(2)['data_interval'][0].split(' ')
# AttributeError: 'Timedelta' object has no attribute 'split'
print(data_merge.head(2)['data_interval'][1].days)
print(data_merge.head(2)['data_interval'][1].seconds)

364
0


In [30]:
data_merge.head(2)['data_interval'].apply(lambda x:x.days)

0    364
1    364
Name: data_interval, dtype: int64

In [49]:
# 转换日期间隔为数字
data_merge['data_interval'] = data_merge['data_interval'].apply(lambda x:x.days)  # 转换成天数
data_merge.head(2)

会员ID         订单号       提交日期    订单金额 max_year_date  data_interval  \
0  15278002468  3000304681 2015-01-01   499.0    2015-12-31            364   
1  39236378972  3000305791 2015-01-01  2588.0    2015-12-31            364   

   year  
0  2015  
1  2015

## 分组聚合

- 按会员ID汇总
- 计算最近一次订单时间
- 计算订单频率
- 计算订单总金额

In [50]:
data_merge.head(2)

会员ID         订单号       提交日期    订单金额 max_year_date  data_interval  \
0  15278002468  3000304681 2015-01-01   499.0    2015-12-31            364   
1  39236378972  3000305791 2015-01-01  2588.0    2015-12-31            364   

   year  
0  2015  
1  2015

In [52]:
rfm = data_merge.groupby(['year','会员ID'],as_index=False).agg({'data_interval':'min', # 最近一次订单时间
                                                             '订单号':'count',  # 频率
                                                             '订单金额':'sum'})  # 总金额
rfm.head()
# 每一年，每个会员，计算最小值间隔日期，订单总数，总金额

year  会员ID  data_interval  订单号    订单金额
0  2015   267            197    2   105.0
1  2015   282            251    1    29.7
2  2015   283            340    1  5398.0
3  2015   343            300    1   118.0
4  2015   525             37    3   213.0

In [51]:
rfm2 = data_merge.groupby(['year','会员ID']).agg({'data_interval':'min', '订单号':'count',  '订单金额':'sum'}) 
rfm2.head()

data_interval  订单号    订单金额
year 会员ID                            
2015 267             197    2   105.0
     282             251    1    29.7
     283             340    1  5398.0
     343             300    1   118.0
     525              37    3   213.0

In [53]:
# 重命名
# R值：值越小，等级越高
# F和M值：值越大，等级越高 
rfm.columns  = ['year','会员ID','r(最近)','f(购买数)','m(总金额)']
rfm.head()

year  会员ID  r(最近)  f(购买数)  m(总金额)
0  2015   267    197       2   105.0
1  2015   282    251       1    29.7
2  2015   283    340       1  5398.0
3  2015   343    300       1   118.0
4  2015   525     37       3   213.0

# 分位数划分RFM区间

In [54]:
rfm.head(2)

year  会员ID  r(最近)  f(购买数)  m(总金额)
0  2015   267    197       2   105.0
1  2015   282    251       1    29.7

In [55]:
rfm.head(2).iloc[:,2:]

r(最近)  f(购买数)  m(总金额)
0    197       2   105.0
1    251       1    29.7

In [56]:
rfm.head(2).iloc[:,2:].describe()

r(最近)    f(购买数)      m(总金额)
count    2.000000  2.000000    2.000000
mean   224.000000  1.500000   67.350000
std     38.183766  0.707107   53.245141
min    197.000000  1.000000   29.700000
25%    210.500000  1.250000   48.525000
50%    224.000000  1.500000   67.350000
75%    237.500000  1.750000   86.175000
max    251.000000  2.000000  105.000000

In [19]:
rfm.iloc[:,2:].describe().T

count         mean          std  min   25%    50%     75%       max
r(最近)   148591.0   165.524043   101.988472  0.0  79.0  156.0   255.0     365.0
f(购买数)  148591.0     1.365002     2.626953  1.0   1.0    1.0     1.0     130.0
m(总金额)  148591.0  1323.741329  3753.906883  1.5  69.0  189.0  1199.0  206251.8

In [21]:
# r最近距离天数 和 m 消费总金额 数据值比较离散 大的很大
# f频次大部分为1次
# 买1次为普通，买2-4次为复购，买5次为高价值
# 如下根据业务规则定义划分依据

r_bins = [-1,79,255,365]
f_bins = [0,2,5,130]
m_bins = [0,69,1199,206251.8]

# RFM表merge等级表

- rfm为每一年每个会员的数据行，加上原有会员等级字段组合表
- merge合并，通过会员ID字段how方式匹配新特征，每行axis0组合

In [24]:
sheet_datas[-1].head(2)

会员ID  会员等级
0       100090     3
1  10012905801     1

In [58]:
rfm.head(2)

year  会员ID  r(最近)  f(购买数)  m(总金额)
0  2015   267    197       2   105.0
1  2015   282    251       1    29.7

In [59]:
rfm_merge = pd.merge(rfm,sheet_datas[-1],on='会员ID',how='inner')
rfm_merge.head(2)

year  会员ID  r(最近)  f(购买数)  m(总金额)  会员等级
0  2015   267    197       2   105.0     1
1  2015   282    251       1    29.7     5

# 特征权重-随机森林

- RandomForestClassifier(随机森林检测每个特征的重要性及每个样例属于哪个类的概率)

https://github.com/morningx/files/blob/master/MachineLearning/%E9%9A%8F%E6%9C%BA%E6%A3%AE%E6%9E%97/%E9%9A%8F%E6%9C%BA%E6%A3%AE%E6%9E%97%E5%8F%98%E9%87%8F%E9%87%8D%E8%A6%81%E6%80%A7-var%E9%87%8D%E8%A6%81%E6%80%A7-%E7%B2%BE%E7%A1%AE%E7%8E%87%E5%8A%A0%E5%99%AA%E9%9F%B3%E5%90%8E%E4%B8%8B%E9%99%8D%E9%87%8F-gini%E9%87%8D%E8%A6%81%E6%80%A7-%E4%B8%8D%E7%BA%AF%E5%BA%A6%E4%B8%8B%E9%99%8D%E7%A8%8B%E5%BA%A6%E4%B9%8B%E5%92%8C.png

In [61]:
# 随机森林树计算字段权重 feature_importances
# random_state随机种子 可固定运行结果
# fit适配X参数和Y标签结果即可得到变量权重
clf = RandomForestClassifier(random_state=0)
clf.fit(rfm_merge[['r(最近)','f(购买数)','m(总金额)']],rfm_merge['会员等级'])

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=0, verbose=0,
                       warm_start=False)

In [62]:
# weights权重；重物；法码
# r最近消费天数 f频次 m金额
# 钱买的多权重高，其次是最近消费天数距离最后样本时间，频次特征不那么重要
weights = clf.feature_importances_
weights
# 间隔天数，频次，金额
# 金额重要性最大，其实是最近购买时间，最差为频次

array([0.41157118, 0.0059908 , 0.58243802])

# RFM正反比关系赋予权重

- array([0.41157118, 0.0059908 , 0.58243802])
- ['r(最近)','f(购买数)','m(总金额)']
- 金额多，则权重高，购买多则权重高，越近值越小间隔天数小权重大
- 对应会员 天数小、金额多、次数多，等级高，用户价值更大
- 天数反比，金额正比，次数正比

- r_bins = [-1,79,255,365]
   f_bins = [0,2,5,130]
   m_bins = [0,69,1199,206251.8]

In [65]:
r_bins = [-1,79,255,365]
f_bins = [0,2,5,130]
m_bins = [0,69,1199,206251.8]

In [66]:
[i for i in range(len(r_bins)-1,0,-1)]

[3, 2, 1]

In [71]:
pd.cut(rfm['r(最近)'],r_bins,labels = [3, 2, 1])

0         2
1         2
2         1
3         1
4         3
         ..
148586    1
148587    2
148588    2
148589    1
148590    2
Name: r(最近), Length: 148591, dtype: category
Categories (3, int64): [3 < 2 < 1]

In [72]:
# RFM分箱得分
rfm['r_socre'] = pd.cut(rfm['r(最近)'],r_bins,labels = [i for i in range(len(r_bins)-1,0,-1)])
rfm.r_socre.value_counts()

2    74059
3    37603
1    36929
Name: r_socre, dtype: int64

In [68]:
[i+1 for i in range(len(f_bins)-1)]

[1, 2, 3]

In [73]:
pd.cut(rfm['f(购买数)'],f_bins,labels = [i+1 for i in range(len(f_bins)-1)])

0         1
1         1
2         1
3         1
4         2
         ..
148586    1
148587    1
148588    1
148589    1
148590    1
Name: f(购买数), Length: 148591, dtype: category
Categories (3, int64): [1 < 2 < 3]

In [77]:
rfm['f_score'] = pd.cut(rfm['f(购买数)'],f_bins,labels = [i+1 for i in range(len(f_bins)-1)])
rfm.f_score.value_counts()

1    142001
2      4534
3      2056
Name: f_score, dtype: int64

In [69]:
[i+1 for i in range(len(m_bins)-1)]

[1, 2, 3]

In [74]:
pd.cut(rfm['m(总金额)'],m_bins,labels = [i+1 for i in range(len(m_bins)-1)])

0         2
1         1
2         3
3         2
4         2
         ..
148586    1
148587    3
148588    3
148589    1
148590    1
Name: m(总金额), Length: 148591, dtype: category
Categories (3, int64): [1 < 2 < 3]

In [75]:
# 计算出每个会员的RFM总得分
rfm['m_score'] = pd.cut(rfm['m(总金额)'],m_bins,labels = [i+1 for i in range(len(m_bins)-1)])
rfm.m_score.value_counts()

2    74191
1    37483
3    36917
Name: m_score, dtype: int64

根据上一节计算出来的RFM因子权重，结合上表的结果，计算出每个会员的RFM总得分：

In [78]:
rfm.head(2)

year  会员ID  r(最近)  f(购买数)  m(总金额) r_socre m_score f_score
0  2015   267    197       2   105.0       2       2       1
1  2015   282    251       1    29.7       2       1       1

In [79]:
rfm.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 148591 entries, 0 to 148590
Data columns (total 8 columns):
year       148591 non-null int64
会员ID       148591 non-null int64
r(最近)      148591 non-null int64
f(购买数)     148591 non-null int64
m(总金额)     148591 non-null float64
r_socre    148591 non-null category
m_score    148591 non-null category
f_score    148591 non-null category
dtypes: category(3), float64(1), int64(4)
memory usage: 7.2 MB


# 总得分=特征权重得分求和

In [83]:
# category分类；种类；范畴 分类值变int值可用于计算
rfm = rfm.apply(np.int32)
rfm.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 148591 entries, 0 to 148590
Data columns (total 8 columns):
year       148591 non-null int32
会员ID       148591 non-null int32
r(最近)      148591 non-null int32
f(购买数)     148591 non-null int32
m(总金额)     148591 non-null int32
r_socre    148591 non-null int32
m_score    148591 non-null int32
f_score    148591 non-null int32
dtypes: int32(8)
memory usage: 5.7 MB


In [86]:
weights[0],weights[1],weights[2]
# r,f,m

(0.4115711783910547, 0.005990804787852282, 0.582438016821093)

In [87]:
rfm['rfm_score'] = rfm['r_socre']*weights[0]+rfm['f_score']*weights[1]+rfm['m_score']*weights[2]
# 随机森林对应的权重值，分箱区间不同对应价值不一样
# 综合样本的整体权重和分箱占比，让更有价值的字段对应的分箱更加明显
rfm.head()

year  会员ID  r(最近)  f(购买数)  m(总金额)  r_socre  m_score  f_score  rfm_score
0  2015   267    197       2     105        2        2        1   1.994009
1  2015   282    251       1      29        2        1        1   1.411571
2  2015   283    340       1    5398        1        3        1   2.164876
3  2015   343    300       1     118        1        2        1   1.582438
4  2015   525     37       3     213        3        2        2   2.411571

In [88]:
rfm.rename(columns={'r_socre':'r_score'},inplace=True) 
rfm.head(2)

year  会员ID  r(最近)  f(购买数)  m(总金额)  r_score  m_score  f_score  rfm_score
0  2015   267    197       2     105        2        2        1   1.994009
1  2015   282    251       1      29        2        1        1   1.411571

# RFM模型分组聚合

In [89]:
rfm_0 = rfm.copy()
rfm_0 = rfm_0.iloc[:,:-1]
# 获取所有行，除了Y标签得分的X特征
rfm_0.head()

year  会员ID  r(最近)  f(购买数)  m(总金额)  r_score  m_score  f_score
0  2015   267    197       2     105        2        2        1
1  2015   282    251       1      29        2        1        1
2  2015   283    340       1    5398        1        3        1
3  2015   343    300       1     118        1        2        1
4  2015   525     37       3     213        3        2        2

In [92]:
rfm_0['r_score'] = rfm_0['r_score'].astype(np.str)
rfm_0['f_score'] = rfm_0['f_score'].astype(np.str)
rfm_0['m_score'] = rfm_0['m_score'].astype(np.str)

In [97]:
rfm_0['r_score'].str.cat(rfm_0['f_score']).str[0]

0         2
1         2
2         1
3         1
4         3
         ..
148586    1
148587    2
148588    2
148589    1
148590    2
Name: r_score, Length: 148591, dtype: object

In [98]:
# 用字符指针实现函数str_cat(s,t),将字符串t复制到字符串s的末端
rfm_0['rfm_group'] = rfm_0['r_score'].str.cat(rfm_0['f_score']).str.cat(rfm_0['m_score'])
rfm_0.head(2)

year  会员ID  r(最近)  f(购买数)  m(总金额) r_score m_score f_score rfm_group
0  2015   267    197       2     105       2       2       1       212
1  2015   282    251       1      29       2       1       1       211

In [46]:
data1 = rfm_0.groupby(['rfm_group'],as_index=False)['year','会员ID'].count()
# data1.columns = ['rfm_group','year','number']
data1.head()

rfm_group   year   会员ID
0       111   9118   9118
1       112  16846  16846
2       113   7535   7535
3       121      9      9
4       122   1015   1015

In [47]:
data2 = rfm_0.groupby(['rfm_group'],as_index=True)['year','会员ID'].count()
# data1.columns = ['rfm_group','year','number']
data2.head()

year   会员ID
rfm_group              
111         9118   9118
112        16846  16846
113         7535   7535
121            9      9
122         1015   1015

In [51]:
data = rfm_0.groupby(['rfm_group','year'],as_index=True)['会员ID'].count()
data.columns = ['rfm_group','year','number']
data.head()

rfm_group  year
111        2015    2180
           2016    1498
           2017    3169
           2018    2271
112        2015    3811
Name: 会员ID, dtype: int64

In [99]:
data = rfm_0.groupby(['rfm_group','year'],as_index=False)['会员ID'].count()
data.columns = ['rfm_group','year','number']
data.head()

rfm_group  year  number
0       111  2015    2180
1       111  2016    1498
2       111  2017    3169
3       111  2018    2271
4       112  2015    3811

In [100]:
data['rfm_group'] = data['rfm_group'].astype(np.int32)
data.head(2)

rfm_group  year  number
0        111  2015    2180
1        111  2016    1498

In [57]:
# data.values
# array([[  111,  2015,  2180],
#       [  111,  2016,  1498],

In [61]:
# data.values.tolist()
# ([[111, 2015, 2180],
#  [111, 2016, 1498],
#  [111, 2017, 3169],

# RFM图形展示

In [101]:
bar3D = Bar3D('RFM组合图',width=900,height=300)

bar3D.add(
    'rfm分组结果',"","",
    data.values.tolist(),   # 将上面表转换成嵌套列表形式
    is_visualmap=True,     # 设置可调节功能
    grid3d_width=200,
    grid3d_height=80,
    grid3d_depth = 80)

bar3D

# 数据透视

In [110]:
rfm_0.shape[0]

148591

In [105]:
rfm_0.columns

Index(['year', '会员ID', 'r(最近)', 'f(购买数)', 'm(总金额)', 'r_score', 'm_score',
       'f_score', 'rfm_group'],
      dtype='object')

In [109]:
rfm_0.groupby('rfm_group')['year'].value_counts()

rfm_group  year
111        2017    3169
           2018    2271
           2015    2180
           2016    1498
112        2017    5737
                   ... 
332        2017       2
333        2018     355
           2017      87
           2016      28
           2015      15
Name: year, Length: 88, dtype: int64

In [111]:
def percent(x):
    return x.shape[0]/rfm_0.shape[0]

percent_data = rfm_0.groupby('rfm_group').apply(percent).sort_values(ascending = False)  # 用户量的占比
cumsum_0 = np.cumsum(percent_data)

In [112]:
cp = pd.DataFrame([percent_data,cumsum_0],index=['用户量占比','用户量的累计占比']).T
cp.index.name = 'RFM分组'
cp.head(2)

用户量占比  用户量的累计占比
RFM分组                    
212    0.247922  0.247922
211    0.128023  0.375945

In [113]:
cp

用户量占比  用户量的累计占比
RFM分组                    
212    0.247922  0.247922
211    0.128023  0.375945
312    0.125546  0.501491
112    0.113372  0.614862
213    0.110168  0.725030
311    0.062413  0.787443
111    0.061363  0.848806
313    0.056134  0.904940
113    0.050710  0.955650
123    0.013002  0.968652
233    0.007039  0.975692
122    0.006831  0.982522
333    0.003264  0.985786
133    0.003177  0.988963
322    0.002759  0.991722
222    0.002517  0.994239
223    0.002490  0.996729
323    0.002463  0.999192
332    0.000249  0.999441
321    0.000229  0.999670
221    0.000162  0.999832
232    0.000087  0.999919
121    0.000061  0.999980
132    0.000013  0.999993
331    0.000007  1.000000